x-pack/filebeat/input/gcs/scheduler.go

```diff
@@ -0,0 +1,253 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+
+	cursor "github.com/elastic/beats/v7/filebeat/input/v2/input-cursor"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/gcs/job"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/gcs/state"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/gcs/types"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-concert/timed"
+)
+
+type scheduler interface {
+	createJobs(objects []*storage.ObjectAttrs, log *logp.Logger) []job.Job
+	Schedule(ctx context.Context) error
+}
+
+type gcsInputScheduler struct {
+	publisher cursor.Publisher
+	bucket    *storage.BucketHandle
+	src       *types.Source
+	cfg       *config
+	state     *state.State
+	log       *logp.Logger
+}
+
+// limiter, is used to limit the number of goroutines from blowing up the stack
+type limiter struct {
+	ctx context.Context
+	wg  sync.WaitGroup
+	// limit specifies the maximum number
+	// of concurrent jobs to perform.
+	limit chan struct{}
+}
+
+// NewGcsInputScheduler, returns a new scheduler instance
+func NewGcsInputScheduler(publisher cursor.Publisher, bucket *storage.BucketHandle, src *types.Source, cfg *config,
+	state *state.State, log *logp.Logger,
+) scheduler {
+	return &gcsInputScheduler{
+		publisher: publisher,
+		bucket:    bucket,
+		src:       src,
+		cfg:       cfg,
+		state:     state,
+		log:       log,
+	}
+}
+
+// Schedule, is responsible for fetching & scheduling jobs using the workerpool model
+func (s *gcsInputScheduler) Schedule(ctx context.Context) error {
+	var pager *iterator.Pager
+	var availableWorkers int
+
+	lmtr := &limiter{limit: make(chan struct{}, s.src.MaxWorkers), ctx: ctx}
+
+	if !s.src.Poll {
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, s.src.BucketTimeOut)
+		defer cancel()
+		for {
+			availableWorkers = s.src.MaxWorkers - len(lmtr.limit)
+			if availableWorkers == 0 {
+				continue
+			} else {
+				break
+			}
+		}
```

Comment:
This calculation is made stale by the loop in `scheduleOnce`. The same is true in the calculation below.

I'm thinking it would be best in the first instance to optimistically assume the availability of `s.src.MaxWorkers` workers and let the runtime sort it out. This will mean that these is sometimes more work than actual workers, but this should not be too much of an issue.

I'm looking at a dynamic work allocator that is more careful than what is done here, but I'm not sure it's necessary right now and the interaction with the approach used in `scheduleOnce` is a little complicated.